### PR TITLE
net: Ensure that origin serialization is consistent

### PR DESCRIPTION
**Pull request by mrobinson**
A recent refactoring (#33531) made a change that resulted in the
`Origin` header including the port even when the default port for a
scheme was used. This made the serialization different from that used
for `rust-url`'s `Origin::ascii_serialization()`, breaking CORS on some
sites. This change makes it so that the serialization is consistent
again.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

### CODE REVIEW

**Comment by mrobinson** [comment]
cc @simonwuelker

**Comment by mrobinson** [comment]
This doesn't seem to fix any WPT tests, but it does fix loading of some real world sites.

File: components/net/tests/http_loader.rs [hunk 1]
```diff
@@ -1457,3 +1458,19 @@ fn test_fetch_compressed_response_update_count() {
...
+        let url = Url::parse(url_string).unwrap();
+        let origin = ImmutableOrigin::new(url.origin());
+        let serialized = format!("{}", serialize_origin(&origin));
+        assert_eq!(serialized, origin.ascii_serialization());
+    };
+
+    ensure_serialiations_match("https://example.com");
+    ensure_serialiations_match("http://example.com");
+    ensure_serialiations_match("https://example.com:1234");
+    ensure_serialiations_match("http://example.com:1234");
```
Comment (mukilan) [inline components/net/tests/http_loader.rs:1474]: Should there be a case for the default port? i.e `ensure_serialiations_match("http://example.com:80");`

File: components/layout_2020/flexbox/layout.rs [hunk 2]
```diff
@@ -1309,7 +1309,7 @@ fn do_initial_flex_line_layout<'items>(
 
 impl InitialFlexLineLayout<'_> {
     fn new<'items>(
-        flex_context: &mut FlexContext,
+        flex_context: &FlexContext,
```
Comment (mukilan) [inline components/layout_2020/flexbox/layout.rs:1312]: Should this be part of this PR?

File: components/net/http_loader.rs [hunk 3]
```diff
@@ -2320,18 +2320,30 @@ fn serialize_request_origin(request: &Request) -> headers::Origin {
...
     // Step 3. Return request’s origin, serialized.
+    serialize_origin(origin)
+}
+
+/// Step 3 of <https://fetch.spec.whatwg.org/#serializing-a-request-origin>.
+pub fn serialize_origin(origin: &ImmutableOrigin) -> headers::Origin {
     match origin {
         ImmutableOrigin::Opaque(_) => headers::Origin::NULL,
         ImmutableOrigin::Tuple(scheme, host, port) => {
+            // Note: This must be kept in sync with `Origin::ascii_serialization()``, which does not
```
Comment (mukilan) [inline components/net/http_loader.rs:2331]: Is this change still in conformance with the spec, or is it expected to be removed once we update the code that uses `ascii_serialization()` (or fixed in rust-url)?

File: components/net/tests/http_loader.rs (hunk 1 above)
Comment (mrobinson) [reply components/net/tests/http_loader.rs:1478]: Good point. I've added this case.

File: components/layout_2020/flexbox/layout.rs (hunk 2 above)
Comment (mrobinson) [reply components/layout_2020/flexbox/layout.rs]: Nope. This wasn't meant to be part of this PR. I've removed it.

File: components/net/http_loader.rs (hunk 3 above)
Comment (mrobinson) [reply components/net/http_loader.rs:2331]: The specifications seem to differ on this. For instance, the original RFC 6454 specification says to omit the port if it's the default port (https://www.rfc-editor.org/rfc/rfc6454#section-6.2) and this is what Chromium does (https://chromium.googlesource.com/chromium/src/+/lkgr/url/scheme_host_port.cc#296). Older versions of the HTML specification had language like this as well (https://www.w3.org/TR/2011/WD-html5-20110525/origin-0.html), but the current one does not (https://html.spec.whatwg.org/#ascii-serialisation-of-an-origin).

It looks like what happens is that the data model of the URL changed such that internally the origin's port can be null when it's the default one (https://github.com/whatwg/html/pull/870). The origin in `rust-url` doesn't have support for this yet (https://github.com/servo/rust-url/issues/821). Once that is fixed this should work properly without an extra check.

For us, the most important thing is that the serialization is consistent, I think.

Comment (mukilan) [reply components/net/http_loader.rs:2331]: That makes sense! Thanks for the additional context!

Comment (mukilan) [inline components/net/http_loader.rs:2331]:
```suggestion
            // Note: This must be kept in sync with `Origin::ascii_serialization()`, which does not
```

**Review by mukilan — APPROVED** [review]

**Comment by mrobinson** [comment]
> Interesting, it's unfortunate that the specs are a bit unclear about this. Thanks for the fix ^^ Though perhaps it would be better to just fix [servo/rust-url#821](https://github.com/servo/rust-url/issues/821), so this change isn't needed anymore?

In parallel, I've also opened servo/rust-url#989 which should allow us to remove this bit of code in the future.

**Pull request merged**